### PR TITLE
docs: add default org-wide Code of Conduct

### PR DESCRIPTION
**Pull request by nejch**
This is similar to #12. The org-level file won't be committed to individual repositories, but will appear on GitHub if the projects don't have their own when this is merged. 

Issue template below:

Default Siemens Code of Conduct
---
Dear project maintainers,

The default Siemens Code of Conduct will be applied to this repository on **October 7, 2024**.

If you'd like to opt out, please [create your own Code of Conduct](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-code-of-conduct-to-your-project).

For more details, see the [pull request in our organization repo](https://github.com/siemens/.github/pull/17).

If you have any further questions feel free to comment in the PR or reach out [via email](mailto:opensource@siemens.com).

### CODE REVIEW

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+## Scope
+
+This Code of Conduct applies within all community spaces, and also applies when
+an individual is officially representing the community in public spaces.
+Examples of representing our community include using an official email address,
+posting via an official social media account, or acting as an appointed
+representative at an online or offline event.
+
+This is the default Code of Conduct for all repositories within the `siemens`
```
Comment (nejch) [inline CODE_OF_CONDUCT.md:59]: I added this with b5a537da08e38d00eba16f166ad2faae9926dfc7 based on the same text in our security policy. Thought it'd fit best in the existing Scope section but not sure if it's buried too deep in here, WDYT @fgreinacher?

**Review by nejch — COMMENTED** [review]
@fgreinacher just a quick question about the scope here. Otherwise, we can maybe go ahead with the issues announcement. Was your idea to have the snippet here e.g. in `scripts/` in this repo or just an internal snippet?

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (fgreinacher) [reply CODE_OF_CONDUCT.md:59]: I think it's perfect here.

Just wondering whether we could get rid of the chapter above "This Code of Conduct applies within all community spaces, [...], or acting as an appointed
representative at an online or offline event.". It feels a bit too vague for me.

Comment (nejch) [reply CODE_OF_CONDUCT.md:53]: Good point, removed that part. It's also less duplicated like this and we already have `If you are unsure whether the Code of Conduct applies feel free to reach out..` below so that should keep things clear.

Comment (fmoessbauer) [reply CODE_OF_CONDUCT.md:53]: > Community leaders have the right and responsibility to remove, edit

Having the right to edit comments is quite problematic, as the edit is not easily recognizable. IMHO. I vote for removing "edit".

Comment (nejch) [reply CODE_OF_CONDUCT.md:53]: @fmoessbauer edits are actually much more visible and auditable. See this thread:

image

I'd keep it as that's the default from the CC CoC and it's also a bit nicer to edit comments than just flat out remove them if they're not completely out of line.

Comment (fmoessbauer) [reply CODE_OF_CONDUCT.md:53]: Ok, but can you also see the old content? If not, I’m still against allowing edits.

Comment (nejch) [reply CODE_OF_CONDUCT.md:53]: Yes, each version is preserved with a little diff. But I'm ok either way, let's see what others say :)

image

Comment (fgreinacher) [reply CODE_OF_CONDUCT.md:53]: I agree that editing is preferable, unless we really want to get fully rid of something (e.g. a leaked secret or a totally misplaced comment).

**Comment by fgreinacher** [comment]
> Was your idea to have the snippet here e.g. in scripts/ in this repo or just an internal snippet?

@nejch Sorry, I missed that question, we can put it here in `scripts/`, no worries. Do you want to do it or should I?

File: scripts/announcements.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    for repo in org.get_repos(type="sources"):
+        contents = (
+            repo.get_contents(".")
+            + get_contents_if_exists(repo, "docs")
+            + get_contents_if_exists(repo, ".github")
+        )
+
+        if existing_file and any([f.path.endswith(existing_file) for f in contents]):
+            continue
```
Comment (fgreinacher) [inline scripts/announcements.py:52]: Should we log an info message in this case?

File: scripts/announcements.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    except KeyError as e:
+        sys.exit(f"Missing environment variable: {e}")
+
+    auth = Auth.Token(token)
+    gh = Github(auth=auth)
+    existing_file = os.getenv("GH_COMMUNITY_FILE")
+
+    org = gh.get_organization(org_name)
+
+    for repo in org.get_repos(type="sources"):
```
[the remark this answers is not in the document]
Comment (nejch) [reply scripts/announcements.py:44]: Good catch, also added this for when `has_issues` is disabled :+1:

**Comment by nejch** [comment]
@fgreinacher added a dry run to the script and a few more changes to catch edge cases with new repos.

I created test issues in `siemens-testing-main` (sorry for the noise if you follow that :grin: ). If you're ok with this we can go ahead and publish to `github.com/siemens`. I thought we'd give people 2 weeks time so they can prep their own CoC if they don't like this one or provide feedback here. Let me know what you think :bow:

**Comment by fgreinacher** [comment]
> @fgreinacher added a dry run to the script and a few more changes to catch edge cases with new repos.
> 
> I created test issues in `siemens-testing-main` (sorry for the noise if you follow that 😁 ). If you're ok with this we can go ahead and publish to `github.com/siemens`. I thought we'd give people 2 weeks time so they can prep their own CoC if they don't like this one or provide feedback here. Let me know what you think 🙇
> 
> Example issue: [siemens-testing-main/first-repo#1](https://github.com/siemens-testing-main/first-repo/issues/1)

@nejch Perfect, feel free to post whenever it works for you 👍

**Review by fgreinacher — APPROVED** [review]

File: CODE_OF_CONDUCT.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+Examples of unacceptable behavior include:
+
+* The use of sexualized language or imagery, and sexual attention or advances of
+  any kind
+* Trolling, insulting or derogatory comments, and personal or political attacks
+* Public or private harassment
+* Publishing others' private information, such as a physical or email address,
+  without their explicit permission
+* Other conduct which could reasonably be considered inappropriate in a
+  professional setting
```
Comment (gernot-h) [inline CODE_OF_CONDUCT.md:37]: Sorry for being a little late here, feel free to ignore, no blockers from my side! I would have two formal points: I would try to sort this list by probability and thus would move the "sexual" point way down as this is nothing I've seen in open source projects so far (in contrast to social media etc.) while e.g. insulting is unfortunately still quite common.

File: CODE_OF_CONDUCT.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  community
+
+Examples of unacceptable behavior include:
+
+* The use of sexualized language or imagery, and sexual attention or advances of
+  any kind
+* Trolling, insulting or derogatory comments, and personal or political attacks
+* Public or private harassment
+* Publishing others' private information, such as a physical or email address,
+  without their explicit permission
```
Comment (gernot-h) [inline CODE_OF_CONDUCT.md:35]:
```suggestion
* Publishing others' private information, such as opinions received over private channels,
  or private data like physical or email address, age, political orientation etc. without their explicit permission
```

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (gernot-h) [reply CODE_OF_CONDUCT.md:53]: Also note that this also applies to commits which you often inevitably edit when resolving conflicts, etc. – something which is also often hard or impossible to track afterwards (when branches are deleted etc.).  So I think this point needs to stay.

File: CODE_OF_CONDUCT.md (hunk 5 above)
Comment (fgreinacher) [reply CODE_OF_CONDUCT.md:35]: I like that a lot 👍

File: CODE_OF_CONDUCT.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+* Accepting responsibility and apologizing to those affected by our mistakes,
+  and learning from the experience
+* Focusing on what is best not just for us as individuals, but for the overall
+  community
+
+Examples of unacceptable behavior include:
+
+* The use of sexualized language or imagery, and sexual attention or advances of
+  any kind
+* Trolling, insulting or derogatory comments, and personal or political attacks
```
Comment (stormc) [inline CODE_OF_CONDUCT.md:32]: Hm, is political attacks meant as targeting an individual? Would it make sense to distinguish misbehavior targeting individuals and groups? Or probably something along the line of 

> Trolling, insulting, derogatory comments, personal and political attacks

as a simple enumeration of unacceptable behavior?

File: CODE_OF_CONDUCT.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+not aligned to this Code of Conduct, and will communicate reasons for moderation
+decisions when appropriate.
+
+## Scope
+
+This is the default Code of Conduct for all repositories within the `siemens`
+organization on GitHub.com.
+
+It does not apply for repositories that have their own Code of Conduct. It also
+does not apply for forks where you should follow the upstream Code of Conduct,
```
Comment (stormc) [inline CODE_OF_CONDUCT.md:57]: Hm, as it's IMO implicit, shouldn't there be an explicit wording about the "direction" of forks?:
1. upstream forks, i.e., forks in `github/siemens` cloned from some upstream source
2. downstream forks, i.e., forks from `github/siemens` repos hosted in some downstream repo

Comment (stormc) [reply CODE_OF_CONDUCT.md:57]: Nitpicking, but it says it doesn't apply to repos having their own CoC, then restating the same for forks, i.e., repos forked into the `github/siemens` namespace, so it's redundant? So, just saying 

> It does not apply for repositories that have their own Code of Conduct.

does include the cases as we cannot control downstream forks of `github/siemens` repos and don't overrule CoCs from repos we fork/clone and host in `github/siemens`. What about the the remaining case: an upstream repo has no CoC but is forked into `github/siemens`?

Comment (nejch) [reply CODE_OF_CONDUCT.md:57]: Regarding downstream forks: we already state this only applies to projects in this namespace only, so that's clear to me it can't apply outside, but I'll try to come up with a better wording.

And for the second case I'd say the same and keep it as is - most forks are meant for contributions back upstream. It'd be a bit presumptuous to automatically apply our own CoC to those even if upstream is missing one. If a fork has a different purpose IMO best to have an explicit file there. WDYT?

File: CODE_OF_CONDUCT.md (hunk 4 above)
Comment (nejch) [reply CODE_OF_CONDUCT.md]: I guess that might be more personal making it less visible in public. But I've reshuffled this a bit, thanks!

**Comment by nejch** [comment]
Thanks everyone for the discussion, I've addressed all comments hopefully with c2a6db22abd0a0aee052b9f55cf4dce0ff05aeb6. Should be ready to go @fgreinacher.

**Comment by fgreinacher** [comment]
Thanks folks for all the feedback!

Here we go 🚀
